### PR TITLE
[ CB ] Return only N results from read_all

### PR DESCRIPTION
**Pull request by iefode**
CVS-146801

### CODE REVIEW

File: src/cpp/src/generation_handle.cpp [hunk 1]
```diff
@@ -46,8 +46,11 @@ std::vector<GenerationOutput> GenerationHandleImpl::read_all() {
         add_partial_result(partial_results, iteration_results);
     }
 
-    for (auto& partial_result: partial_results) {
+    for (auto& partial_result : partial_results) {
         results.push_back(partial_result.second);
+        if (results.size() == m_sampling_params.num_return_sequences) {
+            break;
+        }
```
Comment (mzegla) [inline src/cpp/src/generation_handle.cpp:53]: Results returned by read() and collected in read_all() are unordered. For returning `n` sequences I guess we want those with the highest score right?

**Review by mzegla — COMMENTED** [review]
Could you add a python test for this?

File: src/cpp/src/generation_handle.cpp (hunk 1 above)
Comment (Wovchena) [reply src/cpp/src/generation_handle.cpp:53]: Beams should be sorted within same prompt group
https://github.com/openvinotoolkit/openvino.genai/blob/97a05e1eb733662d639c2fd57bab01c11fa3da27/src/cpp/src/group_beam_searcher.cpp#L462

File: src/cpp/src/generation_handle.cpp [hunk 2]
```diff
@@ -46,8 +46,11 @@ std::vector<GenerationOutput> GenerationHandleImpl::read_all() {
         add_partial_result(partial_results, iteration_results);
     }
 
-    for (auto& partial_result: partial_results) {
+    for (auto& partial_result : partial_results) {
         results.push_back(partial_result.second);
+        if (results.size() == m_sampling_params.num_return_sequences) {
+            break;
```
Comment (ilya-lavrenov) [inline src/cpp/src/generation_handle.cpp:52]: I think we can use std::copy_n here with n = min(num_return_sequences, partial_results.size())

File: src/cpp/src/generation_handle.cpp (hunk 1 above)
Comment (ilya-lavrenov) [reply src/cpp/src/generation_handle.cpp:53]: @Wovchena you point to code which is not related to CB

**Comment by ilya-lavrenov** [comment]
@iefode please, finalize the PR.

File: src/cpp/src/generation_handle.cpp (hunk 1 above)
Comment (Wovchena) [reply src/cpp/src/generation_handle.cpp:53]: That's the point. Beams order should be sorted in CB just like in non CB

**Comment by mzegla** [comment]
I suppose earlier we limited results in `generate()` method. With that change should we remove it from `generate()`?

**Review by ilya-lavrenov — APPROVED** [review]

**Pull request merged**